src/relay/pass/gradient.cc

```diff
@@ -121,21 +121,22 @@ struct ReverseAD : ExprFunctor<ADValue(const Expr &)> {
     return std::make_shared<ADFunction>([this, op_ref](const std::vector<ADValue>& args,
                                                        const Attrs& attrs,
                                                        const tvm::Array<Type>& type_args) {
-        std::vector<Expr> call_args;
-        for (const ADValue& adval : args) {
-          call_args.push_back(adval->get<ADTensor>().foward);
+      std::vector<Expr> call_args;
+      for (const ADValue& adval : args) {
+        call_args.push_back(adval->get<ADTensor>().forward);
+      }
+      auto orig = CallNode::make(op_ref, call_args, attrs, type_args);
```

Comment:
Is it possible to use the real original node instead of a reconstruction? Reconstructing a node may lead to losing some information, e.g. the inferred type `checked_type_`.

Reply:
maybe, but it will require big change in code structure. if such a case come up i will do it.

Reply:
I need `checked_type_` in the integration with the tensor expression ad, mostly for finding out the number of the outputs of the original operation. However, I think I can get this information from other sources. Would passing and reassigning just `checked_type_` be dangerous in this case?

Reply:
@sgrechanik-h can i just rerun type infer? right now every pass will destroy checked_type_ and rebuild from type infer.

Reply:
@MarisaKirisame Not sure what you mean, but rerunning type inference sounds like a bit of an overkill, and I'm not sure it can be done before calling the `FPrimalGradient` attribute. If the `checked_type_` must be reset after running the differentiation pass, then one of the solutions could be setting it before calling `FPrimalGradient` to the original value and then resetting it to nullptr after `FPrimalGradient` has finished, but this feels kinda hacky.

(Also currently I think that in my particular case the proper solution would be to fix the signature of `FTVMCompute` so that it accept input types, not only the out_type. And this is not connected to the automatic differentiation pass.)

Reply:
@sgrechanik-h all pass (FuseOps, AD, ANF, GNF, DeadCodeElimination, FoldScaleAxis) remove the type annotation and rerun it AFAIK. I am not sure why it is an AD-specific issue.

Reply:
@MarisaKirisame I think some passes may benefit from using type information, and, of course, they should use it before erasing it (or recreating the node, I don't think `checked_type_` gets literally erased anywhere). In the case of the code we are currently discussing the node is recreated (and thus type information is erased) before calling to `FPrimalGradient` function which could use type information if it was still there. I don't insist on fixing it if it's difficult or unnatural, because I have only one case where this might be useful, moreover in this single case it would be better to fix a completely different part of Relay.

Reply:
My other passes use type info too. But we just rerun type infer, and we are encoding (rerunning type infer) into pass manager too.